flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java

```diff
@@ -812,29 +812,34 @@ public Collection<SlotInfo> getAllocatedSlotsOf(InstanceID instanceID) {
     // Internal periodic check methods
     // ---------------------------------------------------------------------------------------------
 
-    private void tryReleaseUnusedTaskManagers() {
-        if (checkTaskManagerReleasable()) {
+    private void checkClusterReconciliation() {
+        if (checkResourcesNeedReconcile()) {
             // only declare on needed.
             declareNeededResourcesWithDelay();
         }
     }
 
-    private boolean checkTaskManagerReleasable() {
-        ResourceReleaseResult releaseResult =
-                resourceAllocationStrategy.tryReleaseUnusedResources(taskManagerTracker);
+    private boolean checkResourcesNeedReconcile() {
+        ResourceReconcileResult reconcileResult =
+                resourceAllocationStrategy.tryReconcileClusterResources(taskManagerTracker);
 
-        releaseResult.getPendingTaskManagersToRelease().stream()
+        reconcileResult.getPendingTaskManagersToRelease().stream()
                 .map(PendingTaskManager::getPendingTaskManagerId)
                 .forEach(taskManagerTracker::removePendingTaskManager);
 
-        for (TaskManagerInfo taskManagerToRelease : releaseResult.getTaskManagersToRelease()) {
+        for (TaskManagerInfo taskManagerToRelease : reconcileResult.getTaskManagersToRelease()) {
             if (waitResultConsumedBeforeRelease) {
                 releaseIdleTaskExecutorIfPossible(taskManagerToRelease);
             } else {
                 releaseIdleTaskExecutor(taskManagerToRelease.getInstanceId());
             }
         }
-        return releaseResult.needRelease();
+
+        reconcileResult
+                .getPendingTaskManagersToAdd()
+                .forEach(taskManagerTracker::addPendingTaskManager);
```

Comment:
We should use #allocateResource to ensure the maximum resource limitation is not broken.